### PR TITLE
[FedCM] Add testdriver support for FedCM

### PR DESCRIPTION
**Pull request by tttzach**
Adds testdriver support for Federated Credential Management (FedCM)
Link to spec PR: https://github.com/fedidcg/FedCM/pull/465

### CODE REVIEW

**Comment by tttzach** [comment]
We're under the impression that we wouldn't need an RFC because this is a straightforward implementation of web driver commands but feel free to let us know if that's not the case. Thanks for reviewing!

**Comment by foolip** [comment]
> We're under the impression that we wouldn't need an RFC because this is a straightforward implementation of web driver commands but feel free to let us know if that's not the case. Thanks for reviewing!

That's right, https://fedidcg.github.io/FedCM/#automation is all you need, per the exception in https://github.com/web-platform-tests/rfcs#when-to-use-the-rfc-process.

File: resources/testdriver.js [hunk 1]
```diff
@@ -680,6 +680,134 @@
...
+         *
+         * Matches the `Cancel dialog
+         * <https://fedidcg.github.io/FedCM/#webdriver-canceldialog>`_
+         * WebDriver command.
+         *
+         * @param {WindowProxy} context - Browsing context in which
+         *                                to run the call, or null for the current
+         *                                browsing context.
+         *
+         * @returns {Promise} fulfilled after the cancel request is sent, or rejected
```
Comment (foolip) [inline resources/testdriver.js:695]: Does "after the cancel request is sent" mean that there's no guarantee that the request has been processed when the promise resolves? I suspect that actually we resolve the promise when the request has finished, and if so can you update the wording here to give a stronger guarantee?

File: resources/testdriver.js [hunk 2]
```diff
@@ -680,6 +680,134 @@
...
+         * Matches the `Select account
+         * <https://fedidcg.github.io/FedCM/#webdriver-selectaccount>`_
+         * WebDriver command.
+         *
+         * @param {int} account_index - Index of the account to select.
+         * @param {WindowProxy} context - Browsing context in which
+         *                                to run the call, or null for the current
+         *                                browsing context.
+         *
+         * @returns {Promise} fulfilled after the account selection request is sent,
```
Comment (foolip) [inline resources/testdriver.js:714]: Ditto, is it guaranteed that the selection is done? I'll have the same question for each "request is sent", so not commenting on each of them.

File: resources/testdriver.js [hunk 3]
```diff
@@ -680,6 +680,134 @@
...
+        },
+
+        /**
+         * Selects an account from the Federated Credential Management dialog
+         *
+         * Matches the `Select account
+         * <https://fedidcg.github.io/FedCM/#webdriver-selectaccount>`_
+         * WebDriver command.
+         *
+         * @param {int} account_index - Index of the account to select.
```
Comment (foolip) [inline resources/testdriver.js:709]: Judging by https://jsdoc.app/tags-param.html#names-types-and-descriptions, `int` isn't a supported type in JSDoc. This isn't TypeScript, but FWIW it seems like it's not a TS type either, see https://www.typescriptlang.org/docs/handbook/basic-types.html.

So:

```suggestion
         * @param {number} account_index - Index of the account to select.
```

There isn't any type checking in testdriver.js generally, so I think we'll rely on the WebDriver endpoint returning an error if the input isn't right.

File: tools/wptrunner/wptrunner/executors/actions.py [hunk 4]
```diff
@@ -277,6 +277,85 @@ def __call__(self, payload):
...
+class SelectFedCMAccountAction:
+    name = "select_fedcm_account"
+
+    def __init__(self, logger, protocol):
+        self.logger = logger
+        self.protocol = protocol
+
+    def __call__(self, payload):
+        account_index = payload["account_index"]
+        self.logger.debug("Selecting FedCM account of index: %s" % account_index)
```
Comment (foolip) [inline tools/wptrunner/wptrunner/executors/actions.py:300]: Assuming that a number is passed to the testdriver.js API, `account_index` should be either a Python `int` or `float` now, not a string. But you can avoid this `%s` like this:

```suggestion
        self.logger.debug(f"Selecting FedCM account of index: {account_index}")
```

File: resources/testdriver.js [hunk 5]
```diff
@@ -680,6 +680,134 @@
...
+        },
+
+        /**
+         * Sets whether delay is enabled for the Federated Credential Management dialog
+         *
+         * Matches the `Set delay enabled
+         * <https://fedidcg.github.io/FedCM/#webdriver-setdelayenabled>`_
+         * WebDriver command.
+         *
+         * @param {boolean} enabled - Delay to set.
```
Comment (foolip) [inline resources/testdriver.js:782]: "Delay to set" sounds like it's a duration of time, how about:

```suggestion
         * @param {boolean} enabled - Whether to delay the dialog.
```

File: resources/testdriver.js (hunk 1 above)
Comment (tttzach) [reply resources/testdriver.js]: Agreed that the wording here is kind of iffy - rephrased to give a stronger guarantee. Thanks!

File: resources/testdriver.js (hunk 2 above)
Comment (tttzach) [reply resources/testdriver.js]: Ditto, thank you!

File: resources/testdriver.js (hunk 3 above)
Comment (tttzach) [reply resources/testdriver.js]: Ah fixed, thank you!

File: tools/wptrunner/wptrunner/executors/actions.py (hunk 4 above)
Comment (tttzach) [reply tools/wptrunner/wptrunner/executors/actions.py]: Good point, fixed and thank you!

File: resources/testdriver.js (hunk 5 above)
Comment (tttzach) [reply resources/testdriver.js]: Rephrased, thank you!

**Comment by tttzach** [comment]
Please take another look, @foolip :)

**Review by foolip — APPROVED** [review]

**Pull request merged**